File: daily/2024/1105_Filling_Bookcase_Shelves.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> dp;
    void dfs(vector<vector<int>>& books, int shelfWidth, int index, int curWidth, int curHeight, int totalHeight){
        if (index == books.size()) return;
        int bookWidth = books[index][0];
        int bookHeight = books[index][1];
        // Put on the same level
        if (curWidth + bookWidth <= shelfWidth){
            int AddedHeight = max(0, bookHeight - curHeight);
            if (totalHeight + AddedHeight < dp[index][curWidth + bookWidth]){
                dp[index][curWidth + bookWidth] = totalHeight + AddedHeight;
                dfs(books, shelfWidth, index+1, curWidth+bookWidth, curHeight+AddedHeight, totalHeight+AddedHeight);
            }
        }
        // Put on the next level
        dp[index][bookWidth] = min(dp[index][bookWidth], totalHeight+bookHeight);
        dfs(books, shelfWidth, index+1, bookWidth, bookHeight, totalHeight + bookHeight);
    }

    int minHeightShelves(vector<vector<int>>& books, int shelfWidth) {
        dp = vector<vector<int>>(books.size(), vector<int>(shelfWidth+1, INT_MAX));
        dfs(books, shelfWidth, 0, 0, 0, 0);
        return *min_element(dp[books.size()-1].begin(), dp[books.size()-1].end());
    }
};

int main(){
    vector<vector<int>> books = {{1,1},{2,3},{2,3},{1,1},{1,1},{1,1},{1,2}};
    int shelfWidth = 4;
    // vector<vector<int>> books = {{1,3},{2,4},{3,2}};
    // int shelfWidth = 6;
    Solution S;
    cout << S.minHeightShelves(books, shelfWidth) << endl;
    return 0;
}

File: daily_python/1105_Filling_Bookcase_Shelves.py
import os
import sys
from typing import List
import ast

class Solution:
    def helper(self, index, curWidth, curHeight, totalHeight) -> None:
        if index == len(self.books):
            return
        curBookWidth = self.books[index][0]
        curBookHeight = self.books[index][1]
        if curWidth + curBookWidth <= self.shelfWidth:
            AddedHeight = max(0, curBookHeight - curHeight)
            if (totalHeight + AddedHeight < self.dp[index][curWidth + curBookWidth]):
                self.dp[index][curWidth + curBookWidth] = totalHeight + AddedHeight
                self.helper(index+1, curWidth+curBookWidth, curHeight+AddedHeight, totalHeight+AddedHeight)
        if totalHeight + curBookHeight < self.dp[index][curBookWidth]:
            self.dp[index][curBookWidth] = totalHeight + curBookHeight
            self.helper(index+1, curBookWidth, curBookHeight, totalHeight + curBookHeight)

    def minHeightShelves(self, books: List[List[int]], shelfWidth: int) -> int:
        self.books = books
        self.shelfWidth = shelfWidth
        self.dp = [[float('inf')] * (shelfWidth + 1) for _ in range(len(books))]
        self.helper(0, 0, 0, 0)
        return min(self.dp[-1])

def openFileData(filename: str):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            books = lines[0].split('=')[1].strip().strip(';').replace('{', '[').replace('}',']')
            books = ast.literal_eval(books)
            shelfWidth = int(lines[1].split('=')[1].strip().strip(';'))
            dataset.append((books, shelfWidth))
    return dataset

if __name__ == '__main__':
    if (len(sys.argv) == 1):
        path = os.path.basename(__file__)
        filename = os.path.splitext(path)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = openFileData(filename)
    solution = Solution()
    results = []
    for books, shelfWidth in dataset:
        results.append(solution.minHeightShelves(books, shelfWidth))
    for index, result in enumerate(results):
        print(f'dataset {index+1} : {result}')